### PR TITLE
[PROF-13667] Converter: Verify and enforce 'don't override, only add' behavior

### PR DESCRIPTION
**Pull request by wehzzz**
### What does this PR do?
[PROF-13667](https://datadoghq.atlassian.net/browse/PROF-13667)

Enforce that the host profiler config converters respect user-provided configuration: only add missing keys through inference, never overwrite explicitly set values.

- Add `EnsureDefault` helper that sets a default config value only if the key does not exist or already holds the same value; preserves user overrides
- Update `converter_without_agent.go`: use `EnsureDefault` for `host.arch`, `host.name`, and `os.type` resource attributes in the `resourcedetection` processor. Warn when `host.arch` is disabled
- Update `converter_with_agent.go`: use `EnsureDefault` for `allow_hostname_override` in the `infraattributes` processor. Warn when disabled
- Document the "explicit config wins" behavior in the package doc comment and `comp/host-profiler/README.md`

### Motivation

The converters were unconditionally overwriting user-set leaf configs (e.g., `host.arch: enabled: false` was forced to `true`). Users should be able to explicitly set config values and have them respected, even if they conflict with Datadog's defaults. Incompatible values now generate a warning instead of being silently overwritten.

### Describe how you validated your changes

- Added 6 golden file tests (`preserve-*`) covering edge cases where user config should be preserved:
  - `preserve-host-arch`: user disables `host.arch` with system detector present
  - `preserve-host-name`: user enables `host.name` with system detector present
  - `preserve-os-type`: user enables `os.type` with system detector present
  - `preserve-all-res-attrs`: user sets all three resource attributes to non-default values
  - `preserve-res-attrs-no-system`: user pre-configures resource attributes without system detector in the list
  - `preserve-hostname-override`: user disables `allow_hostname_override` on infraattributes (agent mode)

### Additional Notes

It seems that the `standaloneLogger` (`zap.L().Sugar()`) used in `converter_without_agent.go` isn't setup (`TODO: currently converter helpers use datadog-agent's logger which isn't setup when in Standalone mode`). I'm not sure if I should see some logs or not.

[PROF-13667]: https://datadoghq.atlassian.net/browse/PROF-13667?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: comp/host-profiler/collector/impl/converters/converter_without_agent.go [hunk 1]
```diff
@@ -238,16 +234,18 @@ func (c *converterWithoutAgent) ensureResourceDetectionConfig(resourceDetection
...
 
 	// Only set these defaults if we added the system detector
 	if !hasSystemDetector {
-		if err := Set(resourceDetection, "system::resource_attributes::host.name::enabled", false); err != nil {
+		if _, err := EnsureDefault(resourceDetection, "system::resource_attributes::host.name::enabled", false); err != nil {
 			return err
 		}
-		if err := Set(resourceDetection, "system::resource_attributes::os.type::enabled", false); err != nil {
+		if _, err := EnsureDefault(resourceDetection, "system::resource_attributes::os.type::enabled", false); err != nil {
 			return err
```
Comment (wehzzz) [inline comp/host-profiler/collector/impl/converters/converter_without_agent.go]: Not sure if we want to add some logs here to specify that user-set config mais be conflicting with dd config

File: comp/host-profiler/collector/impl/converters/converters.go [hunk 2]
```diff
@@ -155,19 +169,51 @@ func Set[T any](c confMap, path string, value T) error {
...
 		if !isMap {
-			return fmt.Errorf("path element %q is not a map", key)
+			return nil, "", fmt.Errorf("path element %q is not a map", key)
 		}
 
 		currentMap = childMap
 	}
 
+	return currentMap, target, nil
+}
```
Comment (theomagellan) [inline comp/host-profiler/collector/impl/converters/converters.go]: nit: This feels like the `Ensure` function with extra steps. It could be refactored to use Ensure:
- split `path` into `parent` path and `leaf` name
- use `Ensure` with the `parent` path
- return the result and `leaf` name

What do you think?

File: comp/host-profiler/collector/impl/converters/converters.go [hunk 3]
```diff
@@ -155,19 +169,51 @@ func Set[T any](c confMap, path string, value T) error {
...
 	return nil
 }
 
+// EnsureDefault sets a default value if the key does not exist or already holds the same value.
+// If the key exists with a different value, the existing value is preserved (user override wins).
+// Path segments are separated by "::".
+// Creates intermediate maps as needed.
+// Returns true if the default is active (set or already matching), false if a user override was preserved.
+// Returns an error only if path traversal fails (intermediate element is not a map).
+func EnsureDefault[T any](c confMap, path string, value T) (bool, error) {
```
Comment (theomagellan) [inline comp/host-profiler/collector/impl/converters/converters.go]: nit: I think having `Ensure` that returns a `T` value and `EnsureDefault` a `bool` is confusing.
I think `SetIfMissing` or `SetDefault` would be a better name for it.
What do you think?

File: comp/host-profiler/collector/impl/converters/converter_without_agent.go (hunk 1 above)
Comment (theomagellan) [reply comp/host-profiler/collector/impl/converters/converter_without_agent.go]: These two are not really critical for the Standalone mode. We forced them to `false` because having a `system` detector would enable them by default.
The only necessary one is `host.arch`, as the backend uses the buildID and the host arch to retrieve symbols. I think we should warn the user that disabling `host.arch` will break symbolication

File: comp/host-profiler/collector/impl/converters/converters.go [hunk 4]
```diff
@@ -6,14 +6,29 @@
...
+//
+//   - User-set leaf configs are never overwritten. If a user explicitly sets a value,
+//     even if incompatible with what Datadog needs, the converter preserves it.
+//   - Missing configs are added. If a required leaf config is not defined, the converter
+//     adds it along with all required parent keys using sensible defaults.
+//   - Incompatible values generate warnings. If a user-set value conflicts with
+//     Datadog requirements (e.g. host.arch disabled), a warning is logged but the
+//     user's value is preserved.
+//   - External settings error out. Configs that require external information (API keys,
+//     endpoints) that cannot be inferred will cause an error in standalone mode.
```
Comment (theomagellan) [inline comp/host-profiler/collector/impl/converters/converters.go]: nit: I don't think the package doc should explain the behavior in that much detail given that both converter types (`ConverterWithAgent`, `ConverterWithoutAgent`) already do.
Maybe something more in the lines of
```
// Package converters implements OTEL collector configuration converters for the host profiler.
//
// Converters normalize user-provided OTEL collector configs by adding required Datadog-specific
// components while preserving explicit user configuration values wherever possible.
```
I don't have a strong opinion and the suggestion is Claude's, feel free to ignore if you think it's fine.

File: comp/host-profiler/collector/impl/converters/converters.go (hunk 3 above)
Comment (wehzzz) [reply comp/host-profiler/collector/impl/converters/converters.go]: I'm ok with that, I'll change for `SetDefault`, it would be better

File: comp/host-profiler/collector/impl/converters/converters.go (hunk 4 above)
Comment (wehzzz) [reply comp/host-profiler/collector/impl/converters/converters.go]: I don't have a strong opinion on this. I agree with what you're proposing.

File: comp/host-profiler/collector/impl/converters/converter_without_agent.go (hunk 1 above)
Comment (wehzzz) [reply comp/host-profiler/collector/impl/converters/converter_without_agent.go]: So for these two, we keep SetDefault but no need for additional logging?

File: comp/host-profiler/collector/impl/converters/converter_without_agent.go [hunk 5]
```diff
@@ -238,16 +234,18 @@ func (c *converterWithoutAgent) ensureResourceDetectionConfig(resourceDetection
 	}
 
 	// Always ensure host.arch is enabled
-	if err := Set(resourceDetection, "system::resource_attributes::host.arch::enabled", true); err != nil {
+	if ddDefaultValue, err := SetDefault(resourceDetection, "system::resource_attributes::host.arch::enabled", true); err != nil {
 		return err
+	} else if !ddDefaultValue {
+		standaloneLogger.Warn("host.arch is required but is disabled by user configuration and will break symbol uploading; preserving user value")
```
Comment (theomagellan) [inline comp/host-profiler/collector/impl/converters/converter_without_agent.go]: nit on nit: it's not the uploading that gets broken but the fetch in the backend, I think being less specific would be enough
```suggestion
		standaloneLogger.Warn("host.arch is required but is disabled by user configuration and will break symbolication; preserving user value")
```

**Review by theomagellan — APPROVED** [review]

File: comp/host-profiler/collector/impl/converters/converter_with_agent.go [hunk 6]
```diff
@@ -358,9 +353,10 @@ func (c *converterWithAgent) fixProcessorsPipeline(conf confMap, processorNames
 
 		// Track if we have infraattributes
 		if isComponentType(name, componentTypeInfraAttributes) {
-			// Make sure allow_hostname_override is true
-			if err := Set(processors, name+"::"+fieldAllowHostnameOverride, true); err != nil {
+			if ddDefaultValue, err := SetDefault(processors, name+"::"+fieldAllowHostnameOverride, true); err != nil {
 				return nil, err
+			} else if !ddDefaultValue {
+				log.Warn("allow_hostname_override is required but is disabled by user config; preserving user value")
```
Comment (mackjmr) [inline comp/host-profiler/collector/impl/converters/converter_with_agent.go]: I'm not sure to understand why we are making changes in the converter with agent. We aren't allowing bundled mode users to pass otel config, so this should never happen.

File: comp/host-profiler/collector/impl/converters/converter_without_agent.go [hunk 7]
```diff
@@ -238,16 +234,18 @@ func (c *converterWithoutAgent) ensureResourceDetectionConfig(resourceDetection
 	}
 
 	// Always ensure host.arch is enabled
-	if err := Set(resourceDetection, "system::resource_attributes::host.arch::enabled", true); err != nil {
+	if ddDefaultValue, err := SetDefault(resourceDetection, "system::resource_attributes::host.arch::enabled", true); err != nil {
 		return err
+	} else if !ddDefaultValue {
```
Comment (mackjmr) [inline comp/host-profiler/collector/impl/converters/converter_without_agent.go]: How much would this break the profiles ? I understand not wanting to override user configs, but if it renders the profiles unusable I would rather error out than let a broken full host profiler run.

File: comp/host-profiler/collector/impl/converters/converter_with_agent.go (hunk 6 above)
Comment (theomagellan) [reply comp/host-profiler/collector/impl/converters/converter_with_agent.go]: That's the end goal, the [PR](https://github.com/DataDog/datadog-agent/pull/46331) about not allowing the OTel config to Bundled users is still in draft.
We could've not changed anything in `converter_with_agent` knowing that, but in my opinion it's better for correctness to align every behavior even if we know they are only temporary (especially since the fix is so local).
What do you think?

Comment (mackjmr) [reply comp/host-profiler/collector/impl/converters/converter_with_agent.go]: Sure, we can have this logic temporarily. Thanks for clarifying.

File: comp/host-profiler/collector/impl/converters/converter_without_agent.go (hunk 7 above)
Comment (wehzzz) [reply comp/host-profiler/collector/impl/converters/converter_without_agent.go]: This will break the profile 100% because we need the `buildId` and `host.arch` to retrieve the symbol. We discussed this with @theomagellan and we prefer not to return an error in case the user wants to apply this option. We believe that if the user goes to modify the default configuration, they have a reason to do so, and therefore we prefer to warn them that it will break things without blocking them from making that decision. What do you think? (cc @Gandem )

File: comp/host-profiler/collector/impl/converters/converter_with_agent.go [hunk 8]
```diff
@@ -358,9 +353,10 @@ func (c *converterWithAgent) fixProcessorsPipeline(conf confMap, processorNames
 
 		// Track if we have infraattributes
 		if isComponentType(name, componentTypeInfraAttributes) {
-			// Make sure allow_hostname_override is true
-			if err := Set(processors, name+"::"+fieldAllowHostnameOverride, true); err != nil {
+			if ddDefaultValue, err := SetDefault(processors, name+"::"+fieldAllowHostnameOverride, true); err != nil {
 				return nil, err
+			} else if !ddDefaultValue {
```
Comment (Gandem) [inline comp/host-profiler/collector/impl/converters/converter_with_agent.go]: Nit: no need for `else` here since the other branch returns early.

File: comp/host-profiler/collector/impl/converters/converter_without_agent.go (hunk 7 above)
Comment (Gandem) [inline comp/host-profiler/collector/impl/converters/converter_without_agent.go]: Nit same here, no need for else:
```suggestion
	}
	if !ddDefaultValue {
```

Comment (Gandem) [reply comp/host-profiler/collector/impl/converters/converter_without_agent.go]: This will only break symbolication, i.e. profiles for compiled languages (Go, Rust, C and C++). Profiles for other languages will continue to work.

I'd personally vote for keeping things as is (logging a warning), but making the end user impact more explicit in the warning message (i.e. profiles for compiled languages will be missing symbols).

**Review by Gandem — COMMENTED** [review]
Couple nits, otherwise LGTM 👍

File: comp/host-profiler/collector/impl/converters/converters.go [hunk 9]
```diff
@@ -155,19 +157,51 @@ func Set[T any](c confMap, path string, value T) error {
...
+// Returns true if the default is active (set or already matching), false if a user override was preserved.
+// Returns an error only if path traversal fails (intermediate element is not a map).
+func SetDefault[T any](c confMap, path string, value T) (bool, error) {
+	currentMap, target, err := ensurePath(c, path)
+	if err != nil {
+		return false, err
+	}
+
+	if existingValue, exists := currentMap[target]; exists && !reflect.DeepEqual(existingValue, value) {
+		return false, nil
```
Comment (nsavoire) [inline comp/host-profiler/collector/impl/converters/converters.go]: nit:
```suggestion
	if existingValue, exists := currentMap[target]; exists {
		return reflect.DeepEqual(existingValue, value), nil
```

**Review by nsavoire — APPROVED** [review]
